### PR TITLE
Add CLI tool openqa-validate-yaml

### PR DESCRIPTION
**Pull request by perlpunk**
So one can easily validate local YAML files without having to post to the API.

* Move validation code to OpenQA::JobTemplates
* Add test for invalid schema

Related issue: https://progress.opensuse.org/issues/60329

### CODE REVIEW

File: script/openqa-validate-yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        $data = YAML::XS::LoadFile($file);
+    }
+}
+else {
+    say "Nothing to do";
+    exit;
+}
+
+if ($opt->validate_schema) {
+    say "# Validating Schema also";
```
Comment (Martchus) [inline script/openqa-validate-yaml]: `'# validating schema, too'` or `'# validating also schema'` would sound better to me.

File: script/openqa-validate-yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+my $default_schema = "$Bin/../public/schema/JobTemplates-01.yaml";
+my $schema         = $opt->schema_file // $default_schema;
+unless (-f $schema) {
+    say "Schema file '$schema' does not exist";
+    exit 1;
+}
+my $data;
+
+if ($file) {
```
Comment (Martchus) [inline script/openqa-validate-yaml]: I'd write this as:

```
if (!$file) {
    say "Nothing to do";
    exit;
}

my $data;
```

for less nesting.

File: script/openqa-validate-yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+Validate a YAML file against a schema.
+You can pass a YAML file or '-' for STDIN.
+EOM
+    ['validate-schema', 'Validate Schema file itself also'],
+    ['schema-file=s',   "Schema file to validate against (default public/schema/JobTemplates-01.yaml)"],
+    ['help',            'print usage message and exit', {shortcircuit => 1}],
+);
+
+print($usage->text), exit if $opt->help;
```
Comment (Martchus) [inline script/openqa-validate-yaml]: If this was C/C++ the compiler would emit a warning like "misuse of comma operator". I'm wondering whether this isn't considered bad practice in Perl as well.

File: lib/OpenQA/WebAPI/Plugin/YAMLRenderer.pm [hunk 4]
```diff
@@ -15,8 +15,8 @@
 
 package OpenQA::WebAPI::Plugin::YAMLRenderer;
 use Mojo::Base 'Mojolicious::Plugin';
+use OpenQA::Utils qw(validate_data);
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Plugin/YAMLRenderer.pm]: Maybe use single quotes here for consistency with the line above.

File: script/openqa-validate-yaml (hunk 3 above)
Comment (perlpunk) [reply script/openqa-validate-yaml]: I've taken this directly from  perldoc Getopt::Long::Descriptive.
IMHO using the comma like this is common in perl.

File: lib/OpenQA/WebAPI/Plugin/YAMLRenderer.pm (hunk 4 above)
Comment (perlpunk) [reply lib/OpenQA/WebAPI/Plugin/YAMLRenderer.pm]: done

File: script/openqa-validate-yaml (hunk 2 above)
Comment (perlpunk) [reply script/openqa-validate-yaml]: done

File: script/openqa-validate-yaml (hunk 1 above)
Comment (perlpunk) [reply script/openqa-validate-yaml]: done

File: script/openqa-validate-yaml (hunk 3 above)
Comment (Martchus) [reply script/openqa-validate-yaml]: ok

**Comment by Martchus** [comment]
But please call the additional commit now differently, e.g. 'Improve style of YAML validation code'.

**Comment by perlpunk** [comment]
> But please call the additional commit now differently, e.g. 'Improve style of YAML validation code'.

My plan was to squash all commits in the end anyway...

**Comment by Martchus** [comment]
Seems like the build of `openQA-worker-test` fails now:

```
[   76s] Can't locate JSON/Validator.pm in @INC (you may need to install the JSON::Validator module) (@INC contains: /usr/share/openqa/script/../lib /usr/lib/perl5/site_perl/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/site_perl/5.26.1 /usr/lib/perl5/vendor_perl/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/vendor_perl/5.26.1 /usr/lib/perl5/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/5.26.1 /usr/lib/perl5/site_perl) at /usr/share/openqa/script/../lib/OpenQA/Utils.pm line 36.
```

So likely the new runtime dependency is missing in the worker package.

**Comment by perlpunk** [comment]
> So likely the new runtime dependency is missing in the worker package.

Yeah. Maybe OpenQA::Utils should be split up. The worker itself does not need JSON::Validator...

**Comment by okurz** [comment]
I agree that OpenQA::Utils is now a bit overused. I suggest to move the new functionality elsewhere.

Also I would appreciate if you change the git commit message subject line to resemble the benefit to the user and there definitely *is* benefit when you provide a helpful tool with the new script :)

How about "Add CLI tool to validate a job templates YAML file" as the subject line? I guess reading this from the changelog would be way more helpful than learn about us moving internal functions around.

**Comment by perlpunk** [comment]
I moved the code to `OpenQA::JobTemplates` and adjusted the commit message.

File: openQA.spec [hunk 5]
```diff
@@ -48,7 +48,7 @@
...
+%define main_requires %assetpack_requires git-core perl(Carp::Always) perl(Date::Format) perl(DateTime::Format::Pg) perl(DBD::Pg) >= 3.7.4, perl(DBI) >= 1.632, perl(DBIx::Class) >= 0.082801, perl(DBIx::Class::DeploymentHandler) perl(DBIx::Class::DynamicDefault) perl(DBIx::Class::Schema::Config) perl(DBIx::Class::Storage::Statistics) perl(DBIx::Class::OptimisticLocking) perl(File::Copy::Recursive) perl(Net::OpenID::Consumer) perl(Module::Pluggable) perl(aliased) perl(Config::Tiny) perl(Text::Diff) perl(CommonMark) perl(JSON::Validator) perl(IPC::Run) perl(Archive::Extract) perl(Time::ParseDate) perl(Sort::Versions) perl(BSD::Resource) perl(Pod::POM) perl(Mojo::Pg) perl(Mojo::RabbitMQ::Client) >= 0.2, perl(SQL::Translator) perl(YAML::XS) perl(LWP::UserAgent) perl(Getopt::Long::Descriptive)
```
Comment (okurz) [inline openQA.spec]: If we only need the new dep in a script this should probably be in the next lines on "client_requires"

Comment (perlpunk) [reply openQA.spec]: created https://github.com/os-autoinst/openQA/pull/2683

Comment (perlpunk) [reply openQA.spec]: Had to close #2683

**Pull request merged**